source/Plugins.cpp

```diff
@@ -0,0 +1,133 @@
+/* Plugins.cpp
+Copyright (c) 2022 by Sam Gleske (samrocketman on GitHub)
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Plugins.h"
+
+#include "DataFile.h"
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Files.h"
+
+#include <algorithm>
+#include <map>
+
+using namespace std;
+
+namespace {
+	map<string, bool> settings;
+	map<string, bool> frozenSettings;
+
+	void LoadSettings(const string &path)
+	{
+		DataFile prefs(path);
+		for(const DataNode &node : prefs)
+		{
+			const string &key = node.Token(0);
+			if(key == "plugins")
+				for(const DataNode &child : node)
+					if(child.Size() == 2)
+						settings[child.Token(0)] = child.Value(1);
+		}
+	}
+}
+
+
+
+void Plugins::Load()
+{
+	// Global plugin settings
+	LoadSettings(Files::Resources() + "plugins.txt");
+	// Local plugin settings
+	LoadSettings(Files::Config() + "plugins.txt");
+}
+
+
+
+void Plugins::Save()
+{
+	if(settings.empty())
+		return;
+	DataWriter out(Files::Config() + "plugins.txt");
+
+	out.Write("plugins");
+	out.BeginChild();
+	{
+		for(const auto &it : settings)
+			out.Write(it.first, it.second);
+	}
+	out.EndChild();
+}
+
+
+
+// Freeze the plugin set to determine if a setting has changed.
+void Plugins::Freeze()
+{
+	copy(settings.begin(), settings.end(), inserter(frozenSettings, frozenSettings.end()));
+}
+
+
+
+// Check if a plugin is known. It does not matter if it is enabled or disabled.
+bool Plugins::Has(const string &name)
+{
+	auto it = settings.find(name);
+	return it != settings.end();
+}
+
+
+
+// Determine if a plugin setting has changed since launching.
+bool Plugins::HasChanged(const string &name)
+{
+	return frozenSettings[name] != settings[name];
+}
+
+
+
+// Returns true if any plugin enabled or disabled setting has changed since
+// launched via user preferences.
+bool Plugins::HasChanged()
+{
+	for(const auto &plugin : frozenSettings)
+		if(plugin.second != settings[plugin.first])
```

Comment:
I don’t understand.  Can you elaborate and guide me toward what is good in this case?  Overall purpose of this function is to compare equality of two maps.  So that “restart game” text is displayed to user in pause menu.

Reply:
I think he means you need to use a method to get what the data of this key is, explicitely, one that would be const qualified (and thus wouldnt require a const conversion?)
I dont have a computer rn so idk what method that is

Reply:
I'm still not sure I understand the terminology.  I web searched in order to learn about it but the results were not something I could learn to address this feedback.  So an example of some kind would be great.

Reply:
This method name and signature strongly imply that no changes will be caused by simply invoking it.

However, it uses `operator[]`, which modifies the subject map and inserts a default-constructed value of the appropriate type. If you have a `const map<K, V> foo`, you cannot compile `foo[someKey]` as `foo` is not modifiable.

A const-qualified inclusion check would generally use `map::find`.

Reply:
Thanks for expanding.  I will try to use assignment with the variable being const type.

Reply:
@warp-core helped me through a better const-qualified solution I think.  See b80a33425ba706ee1d12b3bc14dd26935f575a35 which removes frozenSettings and instead uses a pair of bool values.  `map.find` is used instead of `operator[]` accessor.